Improve forms accessibility [MAILPOET-4471]

costasovo opened the pull request:
This PR was [originally submitted](https://github.com/mailpoet/mailpoet/pull/4204) by @simsekb.
I've recreated it because we still have the issue with public builds. I've also added a couple of additional fixes for issues that I missed when doing [the initial review on the original PR](https://github.com/mailpoet/mailpoet/pull/4204).

It changes two things:
1) Improves accessibility of radio buttons, list of segments, and checkboxes by grouping them in a fieldset (see [WAI grouping recommendations](https://www.w3.org/WAI/tutorials/forms/grouping/)). It affects forms that contain a list of segments block, radio buttons custom field block, or checkbox custom field block. It also affects the manage subscription form displayed on the manage subscription page. This PR also adds a CSS reset for fieldsets used inside MailPoets' forms, so the change should not affect the visual appearance of previously created forms.

2) Improves invalid email address error rendering. The error is now rendered next to the field and contains slightly different wording. `Please specify a valid email address.` > `This value should be a valid email.`

[MAILPOET-4471]

[MAILPOET-4471]: https://mailpoet.atlassian.net/browse/MAILPOET-4471?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
Add fieldset and legend to list selection

Made checkboxes in a form accessible according to the WCAG-standards by adding a fieldset around them and adding a legend describing the fieldset's purpose
(name).
[MAILPOET-4471]

Commit:
Improve email error message in forms.

Added a 'type-message' for Parsley to display an error message when the input field is filled in incorrectly.
Also removed the global Parsley message to prevent it from overruling the 'required-message'.
[MAILPOET-4471]

Commit:
Add styling for the for fieldsets to avoid default.

Add styling for the fieldsets to avoid getting the default styling with the black border and unecessary padding and margin.
Applying default or theme styles would cause a BC break for existing forms.
[MAILPOET-4471]

Commit:
Add a fieldset and legend element to the checkboxes.

[MAILPOET-4471]

Commit:
Add a fieldset and legend element to the radio buttons.

[MAILPOET-4471]

costasovo committed:
Change form fieldset CSS reset to also affect manage subscription form

Previously the reset was applied to .mailpoet_form class but manage subscription form
which also uses these form blocks doesn't have the class.
This commit moves the reset to an internal class that wraps every input and is used
also in the manage subscription page.
[MAILPOET-4471]

costasovo committed:
Improve legend tag rendering in forms

The legend html was not escaped and was ignoring block settings coming
from the form editor.
This commit adds a new helper function for rendering legend that respects
applies block settings and also escapes the output.
[MAILPOET-4471]

costasovo committed:
Unify error message for invalid email with other messages

Other error messages are also ended with period.
[MAILPOET-4471]

costasovo committed:
Update unit tests to reflect form rendering changes

[MAILPOET-4471]

costasovo committed:
Update form acceptance test check

[MAILPOET-4471]

triple0t commented on mailpoet/lib/Form/Block/Checkbox.php:
```diff
@@ -70,6 +71,8 @@ public function render(array $block, array $formSettings, ?int $formId = null):
       $html .= '</label>';
     }
 
+    echo '</fieldset>';
+
```
Was this intentional?

costasovo replied:
Nope! Good catch this need's to be fixed.

triple0t commented on mailpoet/assets/css/src/components-public/_public.scss:
```diff
@@ -65,6 +65,15 @@ $form-line-height: 1.4;
 }
 
+/* Reset fieldset styles in form for backward compatibility. */
+.mailpoet_paragraph {
```
Is this class name okay?
This gives the impression that it's a paragraph not a fieldset

costasovo replied:
Yep, it is ok. I agree it is weird, but historically this is a class for wrapping MailPoet form inputs.

triple0t replied:
Okay

triple0t reviewed (commented):
Hi @costasovo 
Thanks for recreating the PR

I'm sorry for the delayed response and Apologies would not be able to review it due to my AFK

I took a quick look and it appears there's an error

Steps:
* Create a new post from the editor  i.e /wp-admin/post-new.php
* Add MailPoet subscription form
* Add a form with Checkbox and Radio custom fields (Create a new form if you don't have one)
* Check the console and notice the error



<img width="1767" alt="Screenshot 2022-07-22 at 1 03 05 PM" src="https://user-images.githubusercontent.com/30554163/180436042-ae749aa4-eacd-41c4-8ef0-d12549c58062.png">

costasovo commented:
>  I took a quick look and it appears there's an error

Thanks for reporting the error. I think this is not related to the task because none of the changes affect the form editor. Actually, it is all about the front-end rendering. I'll try to replicate it.

costasovo committed:
Fix form checkbox html

[MAILPOET-4471]

lysyjan commented on mailpoet/lib/Form/Block/BlockRendererHelper.php:
```diff
@@ -34,7 +34,7 @@ public function getInputValidation(array $block, array $extraRules = [], ?int $f
       $rules['minlength'] = ModelValidator::EMAIL_MIN_LENGTH;
       $rules['maxlength'] = ModelValidator::EMAIL_MAX_LENGTH;
-      $rules['error-message'] = __('Please specify a valid email address.', 'mailpoet');
+      $rules['type-message'] = __('This value should be a valid email', 'mailpoet');
```
I'm not sure about this message. The change makes sense to me, but the text has a slightly different stylization than other messages. But maybe we should address all messages and UX with error messages in a new task. I'm only curious about your opinion.

costasovo replied:
That's a good point about the stylization. On the other hand, I like the new message more :) I don't have a strong opinion here.

lysyjan replied:
Okay. Let's leave it as it is. Thanks

lysyjan commented on mailpoet/assets/css/src/components-public/_public.scss:
```diff
@@ -63,6 +63,12 @@ $form-line-height: 1.4;
     line-height: $form-line-height;
   }
+
+  fieldset {
```
I tried creating a testing form and found a problematic case. I guess it's caused by my specific WP theme ([Storefront](https://pcm.wordpress.org/themes/storefront/)), but our forms could look bad. Here is [an example](https://d.pr/i/Aj81Nc). In my opinion, we should reset the background color for the fieldset and the legend tag. Do you agree?

costasovo replied:
Nice catch! Thank you! I agree.

lysyjan reviewed (commented):
It looks good, but I have two notes about those changes.

costasovo committed:
Improve CSS reset for fieldset and legend in form

These changes address issues found when testing with Storefront theme.
[MAILPOET-4471]

costasovo commented:
@lysyjan I believe I fixed the issue with Storefront theme.

lysyjan reviewed (approved):
👍🏻
